_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,375 @@
+
+```java
+class CityServiceImplTestMockitoDonts {
+
+    // System under Test (SuT)
+    private CityService cityService;
+
+    // Mocks
+    private CityRepository cityRepository;
+
+    // Helper
+    private City expected;
+
+    @BeforeEach
+    void setUp() {
+        expected = createCity();
+        cityRepository = Mockito.mock(CityRepository.class);
+        cityService = new CityServiceImpl(cityRepository);
+
+        // Avoid such complex declarations
+        Mockito.when(cityRepository.save(expected))
+                .thenReturn(Optional.of(expected));
+        Mockito.when(cityRepository.find(expected.getId()))
+                .thenReturn(Optional.of(expected));
+        Mockito.when(cityRepository.findByName(expected.getName()))
+                .thenReturn(Optional.of(expected));
+        Mockito.when(cityRepository.findAllByCanton(expected.getCanton()))
+                .thenReturn(Collections.singleton(expected));
+        Mockito.when(cityRepository.findAllByCountry(expected.getCanton().getCountry()))
+                .thenReturn(Collections.singleton(expected));
+    }
+
+    @Test
+    void save() throws ElementNotFoundException {
+        ReflectionAssert.assertReflectionEquals(expected, cityService.save(expected));
+    }
+
+    @Test
+    void find() throws ElementNotFoundException {
+        ReflectionAssert.assertReflectionEquals(expected, cityService.find(expected.getId()));
+    }
+
+    @Test
+    void delete() throws ElementNotFoundException {
+        cityService.delete(expected);
+        Mockito.verify(cityRepository).delete(expected);
+    }
+
+    @Test
+    void findByName() throws ElementNotFoundException {
+        ReflectionAssert.assertReflectionEquals(expected, cityService.findByName(expected.getName()));
+    }
+
+}
+```
+
+Knowing how to create the mocks, let's have a look at what you should avoid. Otherwise, you test cases are confusing
+what decreases your code's maintainability. It will save you much time debugging and doesn't let your team members guess
+what the intent of the test case is.
+
+* Do not concentrate all the Mockito declarations in the `setUp` method! Even tough the test cases are reduced to a
+  minimum, the readability suffers a lot. Besides, you must highly pay attention to not break any other test cases. Like
+  so, you avoid interfering other tests by overriding the setup.
+* Do not recycle your mocks! It's better to create new mocks for each test case. Otherwise, you might experience
+  unexpected behavior through interfering declarations.
+* Do not expect test cases to be always executed in the same order! Write everything that belongs to your test case into
+  the test method, so that a test case can be executed alone in your IDE or all together within you CI!
+* Do not mock Collections nor value types. It is better to use real instances and values!
+* Do not call `reset()` in a test procedure. This is considered as code smell.
+
+## Mockito Specialities and Restrictions
```

Comment:
The sub sections in this section would fit better into the "Mockito in Use" section above, I would say. I would move the sub sections but keep the questions here and then rename this section to "Mockito FAQ" (I love the question and answer style, but it doesn't really fit to the subsections).